ykrt/src/compile/jitc_yk/codegen/x64/mod.rs

```diff
@@ -419,13 +419,79 @@ impl<'a> Assemble<'a> {
+                        // We assume nothing about the unused upper bits (if any) of `lhs_reg` and
+                        // the `u64` backing the constant, so we can't simply try and use `add r64,
+                        // imm32` to cover lots of cases. This would lead to undesired overflows
+                        // caused by (potentially set) upper bits.
+                        let mut get_lhs_reg = |lhs| {
+                            self.ra.assign_gp_regs(
+                                &mut self.asm,
+                                iidx,
+                                [RegConstraint::InputOutput(lhs)],
+                            )
+                        };
+                        match bit_size {
+                            8 => {
+                                if let Ok(_v8) = i8::try_from(*v as i64) {
+                                    todo!("add r/m8, imm8");
+                                    //return;
+                                }
+                            }
+                            16 => {
+                                if let Ok(_v8) = i8::try_from(*v as i16) {
+                                    todo!("add r/m16, imm8");
+                                    //return;
+                                } else {
+                                    todo!("add r/m16, imm16");
+                                    //return;
+                                }
+                            }
+                            32 => {
+                                if let Ok(v8) = i8::try_from(*v as i32) {
+                                    let [lhs_reg] = get_lhs_reg(lhs.clone());
+                                    dynasm!(self.asm; add Rd(lhs_reg.code()), BYTE v8);
+                                    return;
+                                } else {
+                                    let [lhs_reg] = get_lhs_reg(lhs.clone());
+                                    dynasm!(self.asm; add Rq(lhs_reg.code()), *v as i32);
+                                    return;
+                                }
+                            }
+                            64 => {
+                                if let Ok(v8) = i8::try_from(*v as i64) {
+                                    let [lhs_reg] = get_lhs_reg(lhs.clone());
+                                    dynasm!(self.asm; add Rq(lhs_reg.code()), BYTE v8);
+                                    return;
+                                } else if let Ok(v32) = i32::try_from(*v as i64) {
+                                    let [lhs_reg] = get_lhs_reg(lhs.clone());
+                                    dynasm!(self.asm; add Rq(lhs_reg.code()), v32);
+                                    return;
+                                }
+                            }
+                            _ => (),
+                        }
+                    }
+                    _ => (),
+                }
+
                 let [lhs_reg, rhs_reg] = self.ra.assign_gp_regs(
```

Comment:
Isn't this function `assign_gp_regs` still loading the constants into an extra register via `load_const_into_gp_regs`? That would explain the slowdown: you haven't replaced anything, you just added additional instructions. I may be wrong.

Reply:
Ah, right.